### PR TITLE
feat: Adding payload distribution capabilities to the API

### LINKED ISSUE #16: 📁 Define Structure: Establish payload format with config files, scripts, and binaries.

### PR DESCRIPTION
**Pull request by sysrex**
Adding payload distribution capabilites to the API
- [x] Add Payload
- [x] Deploy path in custom path on remote machine

Closes #16 

## Summary by CodeRabbit

- **New Features**
  - Introduced functionality to associate additional content with instance provisioning, enabling secure delivery of custom payloads.
  - Enhanced the configuration process with robust validation to ensure payload size limits and proper destination paths.
  - Improved reliability with thread-safe operations during content management.

### CODE REVIEW

File: internal/api/v1/routes/routes.go [hunk 1]
```diff
@@ -166,7 +166,7 @@ func BuildURL(routeName string, params map[string]string, queryParams url.Values
 
 	// Replace parameters in the route
```
Comment (MSevey) [inline internal/api/v1/routes/routes.go:167]: can we ask cursor to create a routes_test.go with a table test for BuildURL to make sure there are not other edge cases we are missing?

File: internal/types/infrastructure/validation.go [hunk 2]
```diff
@@ -99,6 +99,25 @@ func (i *InstanceRequest) Validate() error {
 	if i.Image == "" {
 		return fmt.Errorf("image is required")
 	}
+
+	// Validate payload if provided
+	if i.Payload != "" && len(i.Payload) > 1024*1024*100 { // 100MB limit
```
Comment (MSevey) [inline internal/types/infrastructure/validation.go:104]: make a const `MaxPayloadSize`. This number will be useful to export.

File: internal/types/infrastructure/validation.go [hunk 3]
```diff
@@ -99,6 +99,25 @@ func (i *InstanceRequest) Validate() error {
...
+	if i.PayloadPath != "" {
+		// Ensure path starts with / (absolute path)
+		if !strings.HasPrefix(i.PayloadPath, "/") {
+			return fmt.Errorf("payload_path must be an absolute path (start with /)")
+		}
+
+		// Basic path validation - no relative path components
+		if strings.Contains(i.PayloadPath, "..") {
+			return fmt.Errorf("payload_path cannot contain relative path components (..)")
+		}
```
Comment (MSevey) [inline internal/types/infrastructure/validation.go:118]: To help prevent user errors we can also call [Clean](https://pkg.go.dev/path/filepath#Clean) at the API level.

File: internal/compute/ansible.go [hunk 4]
```diff
@@ -26,18 +26,38 @@ type AnsibleConfigurator struct {
 	instances map[string]string
 	// sshKeyPath stores the SSH key path for all instances
 	sshKeyPath string
+	// payloads stores the payloads for all instances
+	payloads map[string]string
+	// payloadPaths stores the payload paths for payloads
+	payloadPaths map[string]string
```
Comment (MSevey) [inline internal/compute/ansible.go:32]:
```suggestion
	// payloads stores the payloads for all instances
	payloads map[string]payload
```

```go
type payload struct {
   name string
   data []byte
   path string
}
```

File: internal/types/infrastructure/provisioning.go [hunk 5]
```diff
@@ -36,6 +39,34 @@ func (i *Infrastructure) RunProvisioning(instances []InstanceInfo) error {
...
+			if instanceName != "" {
+				if provisioner, ok := i.provisioner.(*compute.AnsibleConfigurator); ok {
+					provisioner.SetPayload(instanceName, instanceReq.Payload)
+
+					// Set payload path for the instance
+					provisioner.SetPayloadPath(instanceName, instanceReq.PayloadPath)
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (MSevey) [reply internal/types/infrastructure/provisioning.go:68]: summary, payload data should be added to `InstanceInfo` type, then we can just iterate over instances and not do any checking against i.instances.

**Review by MSevey — CHANGES_REQUESTED** [review]
Summary from Call. 

Base functionality should aim to achieve the following:
- User submits a small payload (i.e. hello world txt file)
- payload is synchronously sent to the instance
- This might require only being able to add payloads to existing instances?? 
- Payload should be very small so can be kept in memory potentially??

To achieve this we should add a safety that the API prevents creating multiple instances if a payload is present. 

Then the next step will be to break the payload distribution off into an asynchronous routine tracked with `Tasks` so that a user can submit larger payloads to multiple instances.

**Comment by MSevey** [comment]
I'll be taking this over.

**Comment by MSevey** [comment]
closing in support of #190
